feat(core/rust/ui): support longer passphrases

grdddj opened the pull request:
Connected with https://github.com/trezor/trezor-firmware/issues/2499:
- enabling longer passphrases up to 50 characters - the limit was already there, just handling all the edge cases and visuals
- also fixing the change of the character categories - previously the swipes were not changing the buttons, as the new buttons were not placed - there is a need to re-place them on each swipe
  - we could solve it also by reusing the 10 already places buttons, and just changing the content of the buttons via `Button::set_content()` - in that case we would need to place them only once in the beginning and would not have to store the `PassphraseKeyboard::key_grid`, which is necessary for later placement

mmilata commented on core/embed/rust/src/ui/model_tt/component/keyboard/passphrase.rs:
```diff
@@ -124,25 +174,16 @@ impl Component for PassphraseKeyboard {
-            });
-            btn.place(area);
-        }
+        self.place_current_buttons();
```
Can you just place all the pages here? IIUC there's no need to redo this after every page change.

grdddj replied:
Thanks for the suggestion, done in [f409b83](https://github.com/trezor/trezor-firmware/pull/2544/commits/f409b8386cbe5fdb4dc4ba93211f802acb213a62)

mmilata reviewed (changes requested):
Mostly looking good, however having 20 as a limit only really works with monospace font. It shouldn't be too difficult to find the longest suffix that fits the textbox (minus length of `...`) using given font.

![emu00000000](https://user-images.githubusercontent.com/85857/194881231-78d36571-5785-4ec6-abb9-5e8159b7eb04.png) ![emu00000001](https://user-images.githubusercontent.com/85857/194882104-df6366cf-7922-4639-8df6-5cf190f76226.png)

grdddj commented:
> Having 20 as a limit only really works with monospace font. It shouldn't be too difficult to find the longest suffix that fits the textbox (minus length of `...`) using given font

Good point, changed in [b94299d](https://github.com/trezor/trezor-firmware/pull/2544/commits/b94299d0ab5eb804ef906278f3023cb53118837e)

In [6209817](https://github.com/trezor/trezor-firmware/pull/2544/commits/62098175a447a8c31e746ced114e990c4bc129b6) I have added some helper macros (copy-paste from `model R` branch) that are helpful during development, and I also already used the `build_string!` to concatenate the ellipsis with the right side of the text

mmilata commented on core/embed/rust/src/ui/macros.rs:
```diff
@@ -4,3 +4,35 @@ macro_rules! include_res {
+            "false"
+        }
+    }};
+}
```
This can be just a regular function as opposed to macro, or am I missing something?

mmilata commented on core/embed/rust/src/ui/model_tt/component/keyboard/passphrase.rs:
```diff
@@ -252,20 +288,53 @@ impl Component for Input {
+                let ellipsis_width = style.font.text_width(ellipsis);
+                let awailable_width_for_text = self.area.width() - ellipsis_width;
+
+                // Finding out how many characters will fit in, starting from the right end.
```
Nit: turning this into a method on `Font` like
```rust
fn longest_suffix(self, width: i16, text: &str) -> &str
// or
fn longest_suffix(self, width: i16, text: &str) -> usize
```
would make this a bit more readable and also potentially reusable.

Also there doesn't seem to be a need to build a string out of this, you can just display the two pieces of text directly?

grdddj replied:
Good idea with creating a `Font` method for this.

With displaying two pieces of text, how can I do it without needing to call the `display::text` two times? (As that would create a lot of extra logic - sometimes calling it once, sometimes twice.)

grdddj replied:
`Font` method done in [6e21a58](https://github.com/trezor/trezor-firmware/pull/2544/commits/6e21a58734413143eb851a2489463dfdc62e8f74)

mmilata replied:
Why do you want to avoid calling it twice? Was thinking of something along the lines of:

```rust
let text_to_display = if fits_whole {
    text
} else {
    let ellipsis_width = ...;
    display_text(text_baseline, "...");
    text_baseline += ellipsis_width; // needed for paint_pending_marker
    text_suffix
}

display::text(text_baseline, text_to_display);
```

grdddj replied:
Tried as an experiment in [2ef1b2a](https://github.com/trezor/trezor-firmware/pull/2544/commits/2ef1b2a7735b778d1a65cd4da17bfeae1b55b64c) - not sure if I like it, as it meant quite a lot of changes. But it removes the need for `heapless::String`, which is certainly helpful

grdddj commented:
I have just noticed that in some cases the very-right pixel of `pending_marker` stays on the screen. It is because it paints itself one pixel longer than the last character, and can therefore be outside of area cleaned by `display::rect_fill`.

We could avoid it by not painting the marker longer than the letter, or by cleaning an area one pixel longer. My preference is probably the first option, as that respects the possible area boundary, and it cannot write into another area (at least not in `x` direction) - it would however make it look not-so-good.

![image](https://user-images.githubusercontent.com/42543243/195558188-78314c74-ccbf-4697-9857-59bc3839b7e1.png)

mmilata commented on core/embed/rust/src/ui/model_tt/component/keyboard/passphrase.rs:
```diff
@@ -285,46 +283,52 @@ impl Component for Input {
+            // Passphrase is longer, showing its right end with ellipsis at the beginning.
+            let ellipsis = "...";
+            let ellipsis_width = style.font.text_width(ellipsis);
+            let awailable_width_for_text = self.area.width() - ellipsis_width;
```
available

grdddj replied:
Fixed in squash

mmilata reviewed (commented):
I don't really understand why we draw the marker 1px longer but judging by the comment it's intentional :shrug:
Otherwise this looks good to me, please drop the unused macros and squash.

grdddj commented:
Force-pushed after rebase on `master`

bosomt commented:
**QA OK**

works as expected /both options - entering via host or via device/

I only noted that when you enter small characters like `,.` via device it looks strange that they dont fit to device display but they only take less than half of it ;) but its expected.

<img width="239" alt="image" src="https://user-images.githubusercontent.com/31506317/197686794-9f040d7d-abce-4094-97ad-6e3301834c17.png">

**Info:**
 - Suite version: desktop 22.10.3 (540db5426ce88c33bdda2fdae0ceb50f6cbc4a23)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/22.10.3 Chrome/100.0.4896.160 Electron/18.3.5 Safari/537.36
 - OS: MacIntel
 - Screen: 1440x900
 - Device: model T 2.5.3 Universal (revision 8668eba9364b0dfb5fa5250a5366421d6ec1dde1)
 - Transport: bridge 2.0.31

grdddj commented:
> 
> <img alt="image" width="239" src="https://user-images.githubusercontent.com/31506317/197686794-9f040d7d-abce-4094-97ad-6e3301834c17.png">
> 

Looking at the photo, it seems to me this is the "legacy" `UI`, and not the "new" `UI2` - visible by the "ellipsis" at the beginning of very long text - old `UI` uses "<" (on photo) but the `UI2` uses "...". Also, we currently do not have a hard limit of characters fitting the screen, we calculate it based on character width.

(Unless you really typed "<" as the first character :))

Did you build the firmware with `UI2=1` option?

bosomt commented:
@grdddj no i just downloaded latest dev master build

maybe that's why i have not seen any change when i tried to test:
```
 -  TT redesign: special screens for altcoins #2500 
 -  each icon (for IconDialog) takes 0.5K of flash, we need to either figure out how to represent three-color image more efficiently than 16bpp, or draw them procedurally #2558 
```

where can i gat correct build that i can upload to device ?

mmilata commented:
Currently we don't even build it in CI. We're planning to switch to UI2 (#2580) shortly _after_ the upcoming release so binaries in the usual places will have it. New & exciting bugs to be expected at that point.

If you really want to test this now here's today's master with UI2=1 (bitcoin-only): [firmware.bin.zip](https://github.com/trezor/trezor-firmware/files/9867489/firmware.bin.zip)

bosomt commented:
@mmilata we will wait after current release. /but i will try that UI2.0 build :) /
Please move all UI related issues to new sprint when freezing current one or creating next one.

<img width="444" alt="image" src="https://user-images.githubusercontent.com/31506317/197989633-54a8fe7e-cdf0-49d0-b6d9-c93945db238f.png">

mmilata commented:
CC @hynek-jina ^

Also for the record UI2 = New UI = Rust UI (noted that this pr has neither in title).
